### PR TITLE
Fix logging of MRI violations log

### PR DESCRIPTION
**Pull request by cmadjar**
This pull request fixes the logic of scan insertion into the `mri_violations_log` table that got compromised with the refactoring of `mri_protocol_checks` (no comma allowed anymore).

### Description of the bug: 

Let's say that for the MINC header `xspace:length` of a T1W, there are multiple `validRange` in the `mri_protocol_checks` like below:
screen shot 2018-07-23 at 3 58 15 pm
For a T1W scan with a `xspace:length` of `176` (fitting the second row in the table above), the insertion pipeline will exclude the T1W scan from insertion and log an entry in the `mri_violations_log` table saying that the `validRange` for a T1W and this header is 10-20 (not taking into consideration the second valid range).
screen shot 2018-07-23 at 4 00 54 pm

### New logic of the pipeline after the fix (validated with Dave on Friday)

1. grep distinct `Header` values in the `mri_protocol_checks` table for the identified `Scan_type`
2. execute the following for both types of severity ("exclude" and "warning")
> a) for each `Header` found in 1., grep all possible values (`validRange` and `validRegex`) that are permitted for this `Scan_type` and `Severity` and check whether the scan is valid based on these values
> b) if the scan has been deemed not valid across all possible values for that `Header`, `Scan_type` and `Severity`, then log the proper information into the `mri_violations_log` table as only one entry. In the example below, there was two entries in the `mri_protocol_checks` for `Header`=`zspace:length`, `Severity`=`warning` and `Scan_type`=`adniT1` (one for `10-20` and one for `30-40`). It inserted the scan into `mri_violations_log` only once with a summary of the possible values for that `Header`, `Severity` and `Scan_type`. 
screen shot 2018-07-23 at 3 39 39 pm
3. behaviour depending on the extra validation status:
> a) if the scan is classified as `exclude`, then the scan will not be inserted into the `files` table and logged into the `mri_violations_log` table and viewable through the MRI violations module.
> b) if the scan is classified as `warn`, then the scan will be inserted into the `files` table with `Caveat` set to `TRUE` and logged into the `mri_violations_log` table (that can be seen through the MRI violations module)
> c) if the scan is classified as `pass`, then the scan will be inserted into the `files` table without any `Caveat` set

### CODE REVIEW

**Comment by MounaSafiHarab** [comment]
@cmadjar 
this has to be  coordinated with the Loris PR as well (namely the mri violated scans module and how it displays the information)...
@llevitis talked to me earlier today.

I just want to make sure we are all on the same page with all these open PRs and what should stay and what should be closed;
I am confuseddddd!

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 1]
```diff
@@ -715,67 +715,219 @@ RETURNS:
...
+        while (my $check = $sth->fetchrow_hashref()) {
+            push(@valid_ranges, $check->{'ValidRange'}) if $check->{'ValidRange'};
+            push(@valid_regexs, $check->{'ValidRegex'}) if $check->{'ValidRegex'};
+        }
+
+        # go to the next header if did not find any checks for that scan
+        # type, severity and header
+        next if (!@valid_ranges && !@valid_regexs);
+
+        # loop through all checks
```
Comment (nicolasbrossard) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: I would reduce this to

```perl
next if grep(NeuroDB::MRI::in_range($value, $_->{'ValidRange'}), @valid_ranges);
```

and if we get past this point, we know that `@valid_ranges` will only contain references to rows for which `$value` is __not__ in range.

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 2]
```diff
@@ -715,67 +715,219 @@ RETURNS:
...
+        my (@failed_valid_ranges, @failed_valid_regexs);
+        foreach my $valid_range (@valid_ranges) {
+            if ($valid_range && (NeuroDB::MRI::in_range($value,
+                $valid_range))) {
+                $is_valid = 1;
+            } else {
+                push(@failed_valid_ranges, $valid_range);
+            }
+        }
+        foreach my $valid_regex (@valid_regexs) {
```
Comment (nicolasbrossard) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Same logic here:

```perl
next if grep($value =~ /$_->{'ValidRegex'}/, @valid_regexs);
```

and if we get past this point, we know that @valid_regexs will only contain references to rows for which `$value` does __not__ match the regex for that row.

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 3]
```diff
@@ -715,67 +715,219 @@ RETURNS:
...
+            } else {
+                push(@failed_valid_regexs, $valid_regex);
             }
+        }
+
+        # go to the next header if the value from the file fits the
+        # value in one of the valid ranges or regex set for that scan type,
+        # header and severity
+        next if $is_valid;
+
```
Comment (nicolasbrossard) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: More concise IMHO

```perl
$valid_fields{$header} = {
    ScanType    => $scan_type,
    HeaderValue => $value,
    ValidRanges => map { $_->{'ValidRange'} } @valid_ranges,
    ValidRegexs => map { $_->{'ValidRegex'} } @valid_regexs
};
```

**Comment by cmadjar** [comment]
> @cmadjarthis has to be coordinated with the Loris PR as well (namely the mri violated scans module and how it displays the information)...
> @llevitis talked to me earlier today.

> I just want to make sure we are all on the same page with all these open PRs and what should stay and what should be closed;
> I am confuseddddd!

@MounaSafiHarab I know it gets confusing. This PR is only taking care of the comma normalization of `mri_protocol_checks` which have been pushed to 20.0 and is currently breaking on the LORIS-MRI front if projects had commas in their tables. The hyphen normalization will be pushed to minor and I will issue a PR to minor for this change when the time comes. Or I will push to your branch @MounaSafiHarab as I think you have a PR for the hyphen normalization of `mri_protocol`.

Does that clarify a bit?

**Comment by cmadjar** [comment]
@nicolasbrossard Thank you very much for the code suggestions!! Looks so much better now :). 

Ready for another round of review when you are ready.

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 4]
```diff
@@ -669,27 +669,27 @@ sub getAcquisitionProtocol {
...
                         $tarchiveInfoRef->{'PatientName'}
                     );
           $message = "\nextra_file_checks from table mri_protocol_check " .
-                     "logged in table mri_violations_log: $checks[0]\n";
+                     "logged in table mri_violations_log: $extra_validation_status\n";
 	  $this->{LOG}->print($message);
 	  # 'warn' and 'exclude' are errors, while 'pass' is not
 	  # log in the notification_spool_table the $Verbose flag accordingly
-	  if (!($checks[0] eq 'pass')){
+	  if (!($extra_validation_status eq 'pass')){
```
Comment (nicolasbrossard) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Nitpick:

```perl
if($extra_validation_status ne 'pass') {
```

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 5]
```diff
@@ -715,67 +715,201 @@ RETURNS:
...
-                        $check->{'Header'},
-                        $value,
-                        $check->{'ValidRange'},
-                        $check->{'ValidRegex'}
-                );
-                push(@faillist, $check->{'ID'});
 
-            }
+    # grep the excluded headers from mri_protocol_check for the scan type
+    my @exclude_headers;
```
Comment (nicolasbrossard) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: I would have done

```perl
$sth->execute($scan_type, 'exclude');
my @exclude_headers = map { $_->{'Header'} } @{ $sth->fetchall_hashref() };
```

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 6]
```diff
@@ -715,67 +715,201 @@ RETURNS:
...
+    $sth->execute($scan_type, 'exclude');
+    while (my $check = $sth->fetchrow_hashref()) {
+        push(@exclude_headers, $check->{'Header'});
+    }
+
+    # grep the warning headers from mri_protocol_check for the scan type
```
Comment (nicolasbrossard) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Same as previous comment

```perl
$sth->execute($scan_type, 'warning');
my @warning_headers = map { $_->{'Header'} } @{ $sth->fetchall_hashref() };
```

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm (hunk 5 above)
Comment (nicolasbrossard) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: I would replace the rest of the code for this function with this

```perl
foreach(qw/exclude warning/) {
    $sth->execute($scan_type, $severity);
    my @headers = map { $_->{'Header'} } @{ $sth->fetchall_arrayref({}) };
    my %validFields = $this->loop_through_protocol_violations_checks(
        $scan_type, $severity, \@headers, $file
    ); 
    if(%validFields) {
        $this->insert_into_mri_violations_log(
            \%validFields, $severity, $pname, $candID, $visit_label, $file
        );
        return $severity;
    }   
}

return 'pass';
```

**Comment by cmadjar** [comment]
@nicolasbrossard Thank you for the suggestions!! All done and ready for another review :)

**Comment by MounaSafiHarab** [comment]
@cmadjar 
tested this and works perfectly!!!!
I tested with dummy data (I used the xpace:length you show in your example) and without your branch, it goes to violated scans logging only the valid range values in the first row of the table (whereas my scan matched the second row);
and with your changes, it goes to the files table (Imaging browser).

I will put passed manual tests but will wait for @nicolasbrossard to test it as well before merging

**Comment by nicolasbrossard** [comment]
I tested this PR "through" `minc_insertion.pl` and it works perfectly. I can see that all lines of `mri_protocol_checks` are taken into account during the extra checks and also that the exclude/warning messages display the concatenation of the valid ranges/regexes.

👍

**Comment by cmadjar** [comment]
@nicolasbrossard Thanks! Changes made :)

**Pull request merged**